### PR TITLE
Fix InsertAwait where LeadingTrivia exists

### LINKED ISSUE #5240: Insert await code fix removes leading blank line
See http://screencast.com/t/BkZjjqLUOVL

### PR DESCRIPTION
**Pull request by diryboy**
Fixes #5240

### CODE REVIEW

**Comment by Pilchie** [comment]
:+1:

@jmarolf?

File: src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs [hunk 1]
```diff
@@ -76,8 +72,19 @@ private bool IsCorrectReturnType(ExpressionSyntax expression, SemanticModel sema
 
         private static ExpressionSyntax ConvertToAwaitExpression(ExpressionSyntax expression)
         {
-            return SyntaxFactory.AwaitExpression(expression)
-                                .WithAdditionalAnnotations(Formatter.Annotation);
+            AwaitExpressionSyntax result;
+
+            if (expression.HasLeadingTrivia)
```
Comment (jmarolf) [inline src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs]: There is no case where we don't want to propagate trivia.  What if there is a comment after this expression?

```C#
return SyntaxFactory.AwaitExpression(expression)
        .WithTriviaFrom(expression)
        .WithAdditionalAnnotations(Formatter.Annotation);
```

File: src/Features/VisualBasic/Portable/CodeFixes/Async/VisualBasicAddAwaitCodeFixProvider.vb [hunk 2]
```diff
@@ -61,7 +58,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeFixes.Async
         End Function
 
         Private Function ConverToAwaitExpression(expression As ExpressionSyntax) As ExpressionSyntax
-            Return SyntaxFactory.AwaitExpression(expression).WithAdditionalAnnotations(Formatter.Annotation)
+            Dim result As AwaitExpressionSyntax
+            If expression.HasLeadingTrivia Then
+                result = SyntaxFactory.AwaitExpression(expression.WithoutLeadingTrivia()).WithLeadingTrivia(expression.GetLeadingTrivia())
```
Comment (jmarolf) [inline src/Features/VisualBasic/Portable/CodeFixes/Async/VisualBasicAddAwaitCodeFixProvider.vb]: Again, should be
```vbnet
Return SyntaxFactory.AwaitExpression(expression)
        .WithTriviaFrom(expression)
        .WithAdditionalAnnotations(Formatter.Annotation)
```

File: src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs (hunk 1 above)
Comment (diryboy) [inline src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs]: @jmarolf Two questions:

1. Shouldn't the first line of your code be `..(expression.WithoutTrivia())`?
1. It seems that our code does not handle adding parenthesis(the only case I can think of that needs propagate the trailing trivia) when needed

For example:
```diff
- return true ? Task.Delay(0) : Task.Delay(1) /* what? */;
+ return await (true ? Task.Delay(0) : Task.Delay(1)) /* what? */;
```

Problems with the above case:

1. VS2015 RTM shows CS4016 but does not provide a fix
1. The comment looks wrongly propagated for me

Comment (jmarolf) [inline src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs]: >"Shouldn't the first line of your code be ..(expression.WithoutTrivia())?"

`WithTriviaFrom(expression)` will override the trivia on that node.

 >For example:
```
- return true ? Task.Delay(0) /* 0 */ : Task.Delay(1) /* 1 */;
+ return await (true ? Task.Delay(0) /* 0 */ : Task.Delay(1)) /* 1 */;
```

I don't understand this case.  This is not going to be valid code.  Are you suggesting we offer a fix here?  What does the containing method for this look like?  

>The comment looks wrongly propagated for me

Handling trivia is difficult.  At the very least we should attempt to keep all the users trivia for them, giving them the ability to fix any sections that get misplaced.

Comment (diryboy) [inline src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs]: I didn't see the overriding code, I remembered it was something like `WithLeadingTrivia(expression.GetLeadingTrivia()).WithTrailingTrivia(expression.GetTrailingTrivia())`. I'll check again later.

Sorry, `Delay` should be `FromResult`, within an `async Task<int> Method`.

Comment (diryboy) [inline src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs]: @jmarolf I created a standalone CodeAnalysis project, placed a breakpoint on the closing curly brace and ran:
```C#
        static void Main( string[] args )
        {
            var tree = CSharpSyntaxTree.ParseText(@"
using System;
using System.Threading.Task;

class Program
{
    async Task<int> A()
    {
        return true ? Task.FromResult(0) /*0*/ : Task.FromResult(1) /*1*/;
    }
}
");
            var root = tree.GetRoot();
            var cond = (ConditionalExpressionSyntax)root.DescendantNodes().First(n => n is ConditionalExpressionSyntax);
            var paren = SyntaxFactory.ParenthesizedExpression(cond).WithTriviaFrom(cond);
        }
```

The syntax on paren looks like `(true ? Task.FromResult(0) /*0*/ : Task.FromResult(1) /*1*/) /*1*/`, I don't think `WithTriviaFrom` did the override.

Comment (jmarolf) [inline src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs]: You are right.  The tokens will be preserved which means we may want something like this:
``` C#
return SyntaxFactory.AwaitExpression(SyntaxFactory.ParenthesizedExpression(expression.WithoutTrivia()))
                                .WithTriviaFrom(expression)
                                .WithAdditionalAnnotations(Formatter.Annotation)
                                .WithAdditionalAnnotations(Simplifier.Annotation);
```

However we will need to add logic to detect the conditional case.  We should special case ConditionalExpressionSyntax and check to make sure both cases are convertable to Task.

Comment (diryboy) [inline src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs]: @jmarolf There are other expression types like this. We need to figure them out first.
For CS4016 cases:
`return null /*0*/ ?? Task.FromResult(1) /*1*/`
`return null /*0*/ as Task<int> /*1*/`
 Can you come up the CS4014 and similar VB cases as well?

**Comment by diryboy** [comment]
@jmarolf Let's discuss the the cases here as it's more visible.

There are other expression types like ConditionalExpressionSyntax. We need to figure them out.

For CS4016 cases:
`return null /*0*/ ?? Task.FromResult(1) /*1*/`
`return null /*0*/ as Task<int> /*1*/`

Can you come up the CS4014 and similar VB cases as well?

**Comment by diryboy** [comment]
retest this please

**Comment by diryboy** [comment]
`Simplifier.Annotation` not works for VB?

**Comment by jmarolf** [comment]
I'm looking into this now.  I'll respond back in a bit.

**Comment by jmarolf** [comment]
Alright, I'm back.  Got distracted on a few other things.

So I merged in your changes and pushed them to [diryboy-FixInsertAwait](https://github.com/jmarolf/roslyn/tree/diryboy-FixInsertAwait)

[AbstractAddAsyncAwaitCodeFixProvider](https://github.com/jmarolf/roslyn/blob/diryboy-FixInsertAwait/src/Features/Core/Portable/CodeFixes/Async/AbstractAddAsyncAwaitCodeFixProvider.cs)
Refactored TryGetTaskAndExpressionTypes into two methods
-  TryGetExpressionType
- TryGetTaskType
```C#
protected static bool TryGetExpressionType(
        SyntaxNode expression,
        SemanticModel semanticModel,
        out INamedTypeSymbol returnType)
{
        var typeInfo = semanticModel.GetTypeInfo(expression);
        returnType = typeInfo.Type as INamedTypeSymbol;
        return returnType != null;
}

protected static bool TryGetTaskType(SemanticModel semanticModel, out INamedTypeSymbol taskType)
{
        var compilation = semanticModel.Compilation;
        taskType = compilation.GetTypeByMetadataName("System.Threading.Tasks.Task");
        return taskType != null;
}
```
Using `semanticModel.GetTypeInfo(expression)` is going to solve all our problems in terms of getting the named type.  No special casing ternary operators.

[CSharpAddAwaitCodeFixProvider](https://github.com/jmarolf/roslyn/blob/diryboy-FixInsertAwait/src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs) and [VisualBasicAddAwaitCodeFixProvider](https://github.com/jmarolf/roslyn/blob/diryboy-FixInsertAwait/src/Features/VisualBasic/Portable/CodeFixes/Async/VisualBasicAddAwaitCodeFixProvider.vb) now have methods that check whether they need to enclose the expression in parenthesis.
```C#
private static SyntaxNode ConvertToAwaitExpression(ExpressionSyntax expression)
{
    var root = expression.Ancestors().Last();
    if (RequiresParenthesis(expression, root))
    {
        expression = expression.Parenthesize();
    }

    return SyntaxFactory.AwaitExpression(expression.WithoutTrivia())
        .WithTriviaFrom(expression)
        .WithAdditionalAnnotations(Formatter.Annotation)
        .WithAdditionalAnnotations(Simplifier.Annotation);
}

private static bool RequiresParenthesis(ExpressionSyntax expression, SyntaxNode root)
{
    var parenthesizedExpression = SyntaxFactory.AwaitExpression(SyntaxFactory.ParenthesizedExpression(expression));
    var newRoot = root.ReplaceNode(expression, parenthesizedExpression);
    var newNode = newRoot.FindNode(expression.Span)
        .DescendantNodesAndSelf(n => n.Kind() != SyntaxKind.ParenthesizedExpression)
        .OfType<ParenthesizedExpressionSyntax>().FirstOrDefault();
    if (newNode != null)
    {
        return !newNode.CanRemoveParentheses();
    }

    return false;
}
```

I'll be tracking down why the simplifier doesn't choose to remove unnecessary parenthesis, in the mean-time I would appreciate if you took a look at these suggestions and incorporated them where appropriate.

**Comment by mmitche** [comment]
Can one of the admins approve this PR test or whitelist this user?

**Comment by diryboy** [comment]
Thanks for the update, @jmarolf I'll pull that down and take a look

**Comment by diryboy** [comment]
@jmarolf I think it's ready for review.

To summarize my changes:

1. Merged most of your changes, except the C# version of ConvertToAwaitExpression, see b7fee1d
1. Tighten the `Parenthesize()` methods a little bit, see 8ae2c82
1. Special case handle single line ternary expressions with trailing trivia, see 690c9ab
1. Various incorrectly passed test cases

**Comment by diryboy** [comment]
retest this please

**Comment by jmarolf** [comment]
test full please

**Comment by diryboy** [comment]
retest this please

**Comment by dpoeschl** [comment]
Just investigated with @jmarolf and @amcasey, the parenthesization tests should be failing but we had an existing ```public void Test()``` method in there that caused xUnit to not run any tests in the entire type. :cry:

@jmarolf Is fixing that up and will rerun your PR validation checks, @diryboy, at which point you should be able to sync and test locally.

**Comment by diryboy** [comment]
Thanks for the update @dpoeschl, I also noticed that `Test()` method but did not raise any suspicion. I'll change that name and fix if any tests actually failed.

**Comment by jmarolf** [comment]
please test all

**Comment by jmarolf** [comment]
restest all please

**Comment by jmarolf** [comment]
retest win please

**Comment by jmarolf** [comment]
retest windows please

**Comment by jmarolf** [comment]
retest this please

**Comment by jmarolf** [comment]
@diryboy I've fixed the test infastructure issue.  Well just need to fix the parenthesis issue and then we can merge this change in.

**Comment by jmarolf** [comment]
@diryboy Look good, merging this in.

**Pull request merged**